driver-core/src/main/com/mongodb/internal/operation/AsyncQueryBatchCursor.java

```diff
@@ -98,7 +98,6 @@
         this.count.addAndGet(firstBatch.getResults().size());
         if (result != null) {
             this.operationTime = result.getTimestamp(OPERATION_TIME, null);
-            this.postBatchResumeToken = getPostBatchResumeTokenFromResponse(result);
```

Comment:
Also, was a similar change not required in the sync code? I think the `QueryBatchCursor` constructor caches the post batch resume token because it calls `initFromCommandResult`.

Reply:
Tgee `QueryBatchCursor` constructor actually calls `initFromQueryResult`, which doesn't cache the post batch resume token.

Reply:
Just had a peek at the [specification](https://github.com/mongodb/specifications/blob/master/source/change-streams/change-streams.rst#response-format), and it indicates that the response from the aggregate command, not just the getMore command, can contain a `postBatchResumeToken`.  Given that, I think the bug in in the sync code, and the async code is actually correct.

Reply:
If I'm correct, then we should revert this change, and open a bug report, because we'll want to backport the fix to 4.0 and 3.12 releases

Reply:
This is based of a very cursory reading, but my understanding is this:

1. `AsyncQueryBatchCursor` constructor always caches the PBRT in the `aggregate` response if it exists.
2. `QueryBatchCursor` never caches the PBRT in the `aggregate` response.

If this is correct, I think there are bugs in both. Per https://github.com/mongodb/specifications/blob/master/source/change-streams/change-streams.rst#updating-the-cached-resume-token, the PBRT should be cached if:

1. The cursor batch is empty. This allows the change stream to continuously update the resume token even if there are no documents returned.
2. The user has iterated to the last document in a non-empty batch (i.e. the next call to `Next` will do a `getMore`).

The async code is caching the resume token even though the batch is potentially non-empty. This means the BatchCursor's resume token might be further ahead in the future than the correct resume token, which would cause documents to be skipped if it were used to create a resumed change stream.

Reply:
I have reverted this change, and I will create a ticket to investigate the sync side.